sdk/helper/random/string_generator.go

```diff
@@ -0,0 +1,261 @@
+type StringGenerator struct {
+	// Length of the string to generate.
+	Length int `mapstructure:"length" json:"length"`
+
+	// Rules the generated strings must adhere to.
+	Rules serializableRules `mapstructure:"-" json:"rule"` // This is "rule" in JSON so it matches the HCL property type
+
+	// Charset to choose runes from. This is computed from the rules, not directly configurable
+	charset runes
+
+	// rng for testing purposes to ensure error handling from the crypto/rand package is working properly.
+	rng io.Reader
+}
+
+// Generate a random string from the charset and adhering to the provided rules.
+func (g *StringGenerator) Generate(ctx context.Context) (str string, err error) {
+	if _, hasTimeout := ctx.Deadline(); !hasTimeout {
+		var cancel func()
+		ctx, cancel = context.WithTimeout(ctx, 1*time.Second) // Ensure there's a timeout on the context
+		defer cancel()
+	}
+
+	// Ensure the generator is configured well since it may be manually created rather than parsed from HCL
+	err = g.validateConfig()
+	if err != nil {
+		return "", err
+	}
+
+LOOP:
+	for {
+		select {
+		case <-ctx.Done():
+			return "", fmt.Errorf("timed out generating string")
+		default:
+			str, err = g.generate()
+			if err != nil {
+				return "", err
+			}
+			if str == "" {
+				continue LOOP
+			}
+			return str, err
+		}
+	}
+}
+
+func (g *StringGenerator) generate() (str string, err error) {
+	// If performance improvements need to be made, this can be changed to read a batch of
+	// potential strings at once rather than one at a time. This will significantly
+	// improve performance, but at the cost of added complexity.
+	candidate, err := randomRunes(g.rng, g.charset, g.Length)
+	if err != nil {
+		return "", fmt.Errorf("unable to generate random characters: %w", err)
+	}
+
+	for _, rule := range g.Rules {
+		if !rule.Pass(candidate) {
+			return "", nil
+		}
+	}
+
+	// Passed all rules
+	return string(candidate), nil
+}
+
+// randomRunes creates a random string based on the provided charset. The charset is limited to 255 characters, but
+// could be expanded if needed. Expanding the maximum charset size will decrease performance because it will need to
+// combine bytes into a larger integer using binary.BigEndian.Uint16() function.
+func randomRunes(rng io.Reader, charset []rune, length int) (candidate []rune, err error) {
+	if len(charset) == 0 {
+		return nil, fmt.Errorf("no charset specified")
+	}
+	if len(charset) > math.MaxUint8 {
+		return nil, fmt.Errorf("charset is too long: limited to %d characters", math.MaxUint8)
+	}
+	if length <= 0 {
+		return nil, fmt.Errorf("unable to generate a zero or negative length runeset")
+	}
+
+	if rng == nil {
+		rng = rand.Reader
+	}
+
+	charsetLen := byte(len(charset))
+	data := make([]byte, length)
+	_, err = rng.Read(data)
+	if err != nil {
+		return nil, err
+	}
+
+	runes := make([]rune, 0, length)
+	for i := 0; i < len(data); i++ {
+		r := charset[data[i]%charsetLen]
```

Comment:
I don't _believe_ it's a problem, but do we run into any problems with the use of modulus here? I know this keeps us within the length of the available charset - wasn't sure if there were any side effects I may not have thought of.

Also, I assume that the RNG in use here will take advantage of an HSM if it's present, using the entropy augmentation stuff?

Reply:
There actually is a problem using the modulus. Not technically with the mod operation itself, but if we aren't careful we can create bias towards the front of the charset. I've put in a fix for this and confirmed that bias has been removed (at least bias that I can see when testing it). I put in a rudimentary bias test as well that runs the function many times, then counts the occurrence of each character and ensures that the standard deviation of each character count is sufficiently small.

When discussing it, @xntrik found this excellent description of the bias problem: [How "random" are the generated passwords?](https://discussions.agilebits.com/discussion/23842/how-random-are-the-generated-passwords)